src/lib/SaferNames/Name.hs

```diff
@@ -0,0 +1,478 @@
+-- Copyright 2021 Google LLC
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE QuantifiedConstraints #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE PolyKinds #-}
+
+module SaferNames.Name (
+  S (..), RawName, Name (..), Env  (..), (<>>), (<.>),
+  emptyEnv, envLookup, envAsScope, withFresh,
+  PlainBinder (..), Scope, RenameEnv, FreshExt, NameTraversal (..),
+  extendNameTraversal, injectNameTraversal, extendInjectNameTraversal,
+  RenameTraversal, voidScopeEnv,
+  E, B, HasNamesE (..), HasNamesB (..), unsafeCoerceE, unsafeCoerceB,
+  NameSet, fmapNames, foldMapNames, freeNames, (@>), (@@>),
+  freshenAbs, freshenBinder, extendRecEnv,
+  projectName, injectNameL, injectNameR, projectNamesL, injectNamesL, injectEnvNamesL,
+  Abs (..), FreshAbs (..), Nest (..), NestPair (..),PlainBinderList, envMapWithKey,
+  AnnBinderP (..), AnnBinderListP (..), EnvE (..), RecEnv (..), RecEnvFrag (..),
+  AlphaEq (..), UnitE (..), VoidE, EmptyNest, PairE (..), MaybeE (..), ListE (..),
+  EitherE (..), LiftE (..), emptyFreshExt, composeFreshExt, idRenamer,
+  EqE, EqB, FreshBinder (..), lookupNameTraversal, extendScope) where
+
+import Control.Monad.Identity
+import Control.Monad.Writer.Strict
+import qualified Data.Set        as S
+import Unsafe.Coerce
+import Data.Text.Prettyprint.Doc  hiding (nest)
+import GHC.Exts (Constraint)
+
+import qualified SaferNames.LazyMap as LM
+
+import qualified Env as D
+
+data S = (:=>:) S S
+       | UnsafeMakeS
+       | UnitScope
+       | VoidScope
+
+-- TODO: we reuse the old `Name` to make use of the GlobalName name space while
+-- we're using both the old and new systems together.
+-- TODO: something like this instead:
+--    type Tag = T.Text
+--    data RawName = RawName Tag Int deriving (Show, Eq, Ord)
+type RawName = D.Name
+
+-- invariant: the raw name in `Name s` is contained in the list in the `ScopeVal s`
+newtype Name (n::S) = UnsafeMakeName RawName
+                      deriving (Show, Eq, Ord)
+
+newtype Env (n::S) (a:: *) = UnsafeMakeEnv (LM.LazyMap RawName a)
+                             deriving (Functor, Foldable, Traversable)
+type Scope n = Env n ()
+
+emptyEnv :: Env (n:=>:n) a
+emptyEnv = UnsafeMakeEnv mempty
+
+voidScopeEnv :: Env VoidScope a
+voidScopeEnv = UnsafeMakeEnv mempty
```

Comment:
Just that they have different types. `Scope VoidScope` and `Scope (n:=>:n)` are both empty, but their types are not considered equal.

Reply:
(It's not important that they have different types, it's just works out that way because of our representation. It would be better if we could have a single type for the empty set but it's harder to arrange.)